<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [50]:
import pandas as pd
import psycopg2
from bs4 import BeautifulSoup
import requests
import warnings
warnings.filterwarnings('ignore')

In [51]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [52]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [53]:
# текст запроса
query_3_1 = f'''
SELECT 
    COUNT(id) as count
    FROM public.vacancies
'''

In [54]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
print('Количество вакансий в БД:', df_3_1.iloc [0]['count'])

Количество вакансий в БД: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [55]:
# текст запроса
query_3_2 = f'''
SELECT 
    COUNT(id) as count
FROM public.employers
'''

In [56]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
print('Количество работодателей в БД:', df_3_2.iloc [0]['count'])

Количество работодателей в БД: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [57]:
# текст запроса
query_3_3 = f'''
SELECT 
    COUNT (id) as count
FROM public.AREAS
'''

In [58]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
print('Количество регионов в БД:', df_3_3.iloc [0]['count'])

Количество регионов в БД: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [59]:
# текст запроса
query_3_4 = f'''
SELECT 
    COUNT (id) as count
FROM public.INDUSTRIES
'''

In [60]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
print('Количество сфер деятельности в БД:', df_3_4.iloc [0]['count'])

Количество сфер деятельности в БД: 294


***

### **Вывод:** 
В данной базе содержится 49197 вакансий, 23501 работодателей, из 1362 Регионов, и охватывают 294 сферы деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [61]:
# текст запроса
query_4_1 = f''' 
SELECT
    a.name as "Регион",
    COUNT(v.name) as "Кол-во вакансий"
FROM public.VACANCIES v
JOIN public.AREAS a on v.area_id = a.id
GROUP BY 1
ORDER BY 2 desc
LIMIT 10
''' 

In [62]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
display(df_4_1)

Регион  Кол-во вакансий
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892
5     Екатеринбург             1698
6  Нижний Новгород             1670
7           Казань             1415
8        Краснодар             1301
9           Самара             1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [63]:
# текст запроса
query_4_2 = f''' 
SELECT
    COUNT(v.name) as count
FROM public.VACANCIES v
WHERE salary_from is not null or salary_to is not null
'''

In [64]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
print('Количество вакансий где заполнено хотя бы одно из двух полей с зарплатой:', df_4_2.iloc [0]['count'] )

Количество вакансий где заполнено хотя бы одно из двух полей с зарплатой: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [65]:
# текст запроса
query_4_3 = f''' 
SELECT
    ROUND(avg(v.salary_from)) as "средние нижней границы",
    ROUND(avg(v.salary_to)) as "средние верхней границы"
FROM public.VACANCIES v
'''

In [66]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
display(df_4_3)

средние нижней границы  средние верхней границы
0                 71065.0                 110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [67]:

# текст запроса
query_4_4 = f''' 
SELECT
    v.schedule as "тип рабочего графика",
    v.employment as "тип трудоустройства",
    COUNT(v.name) as count
FROM public.VACANCIES v
GROUP BY 1, 2
ORDER BY 3 desc
'''

In [68]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
display(df_4_4)

тип рабочего графика  тип трудоустройства  count
0           Полный день     Полная занятость  35367
1      Удаленная работа     Полная занятость   7802
2         Гибкий график     Полная занятость   1593
3      Удаленная работа  Частичная занятость   1312
4        Сменный график     Полная занятость    940
5           Полный день           Стажировка    569
6        Вахтовый метод     Полная занятость    367
7           Полный день  Частичная занятость    347
8         Гибкий график  Частичная занятость    312
9           Полный день     Проектная работа    141
10     Удаленная работа     Проектная работа    133
11        Гибкий график           Стажировка    116
12       Сменный график  Частичная занятость    101
13     Удаленная работа           Стажировка     64
14        Гибкий график     Проектная работа     18
15       Сменный график           Стажировка     12
16       Вахтовый метод     Проектная работа      2
17       Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [69]:
# текст запроса
query_4_5 = f''' 
SELECT
    v.experience as "Требуемый опыт работы",
    COUNT(v.name) as "Колличество вакансий"
FROM public.VACANCIES v
GROUP BY 1
ORDER BY 2 
'''

In [70]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
display(df_4_5)

Требуемый опыт работы  Колличество вакансий
0           Более 6 лет                  1337
1             Нет опыта                  7197
2         От 3 до 6 лет                 14511
3    От 1 года до 3 лет                 26152

***

## **Вывод:**
* В пятерку городов лидеров по колличеству вакансий входят Москва, Санкт-Питербург, Минск, Новосибирск, Алмааты Это самые крупные региональные центры СНГ.

* 24073 вакансии имеют верхнюю или нижнюю границу зарплаты, это означает что в половине вакансий не указанна желаемая зарплата.

* Средняя нижняя граница равна 71065 руб, а средняя верхняя 110537 руб.

* 1  по популярности график работы и тип трудоустройства является (Полный день - Полная занятость),  2 по популярности  (Удаленная работа - Полная занятость).

* По колличеству вакансий от требуемого опыта вакансии распределены следующим образом Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет что показывает что большинство вакансий имеют требование опыта работы От 1 года до 3 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [71]:
# текст запроса
query_5_1 = f''' 
SELECT
    e.name as "Работодатели",
    COUNT(v.name) as "Колличество вакансий"
FROM public.VACANCIES v
JOIN public.employers e on v.employer_id = e.id
GROUP BY 1
ORDER BY 2 desc
LIMIT 5 
'''

In [72]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
display(df_5_1)
print('На 1 месте по количеству вакансий находится:', df_5_1.iloc [0][0],',', 'на 5 по количеству вакансий находится', df_5_1.iloc [4][0],'.')

Работодатели  Колличество вакансий
0         Яндекс                  1933
1     Ростелеком                   491
2       Тинькофф                   444
3           СБЕР                   428
4  Газпром нефть                   331

На 1 месте по количеству вакансий находится: Яндекс , на 5 по количеству вакансий находится Газпром нефть .


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [73]:
# текст запроса
query_5_2 = f''' 
SELECT
    a.name as "Регион",
    COUNT(e.id) as "Колличество работодателей",
    COUNT(v.id) as "Колличество вакансий"
FROM public.AREAS a
LEFT JOIN public.employers e on e.area = a.id
LEFT JOIN public.vacancies v on v.area_id = a.id
GROUP BY 1
HAVING count(v.id) = 0
ORDER BY 2 desc
LIMIT 10 
'''

In [74]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
display(df_5_2)
print('Регион в котором нет вакансий и наибольшее количество работодателей:', df_5_2.iloc [0][0])

Регион  Колличество работодателей  Колличество вакансий
0                 Россия                        410                     0
1              Казахстан                        207                     0
2     Московская область                         75                     0
3     Краснодарский край                         19                     0
4     Ростовская область                         18                     0
5               Беларусь                         18                     0
6            Азербайджан                         17                     0
7  Нижегородская область                         16                     0
8   Республика Татарстан                         16                     0
9             Узбекистан                         15                     0

Регион в котором нет вакансий и наибольшее количество работодателей: Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [75]:
# текст запроса
query_5_3 = f''' 
SELECT
    e.name as "Работодатели",
    COUNT(distinct v.area_id) as "Колличество вакансий"
FROM public.employers e
JOIN public.vacancies v on e.id = v.employer_id
GROUP BY 1
ORDER BY 2 desc
LIMIT 10 
'''

In [76]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
display(df_5_3)

Работодатели  Колличество вакансий
0                      Яндекс                   181
1                  Ростелеком                   152
2                  Спецремонт                   116
3      Поляков Денис Иванович                    88
4                   ООО ЕФИН                     71
5                  Совкомбанк                    63
6                         МТС                    55
7  ЭФКО, Управляющая компания                    49
8                Почта России                    48
9                        КРОН                    48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [77]:
# текст запроса
query_5_4 = f'''
SELECT
    COUNT(e.id) as "Колличество работодателей"
FROM public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL
'''

In [78]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
print('Количество работодателей у которые отсутствует сфера деятельности:', df_5_4.iloc [0][0])

Количество работодателей у которые отсутствует сфера деятельности: 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [79]:
# текст запроса
query_5_5 = f'''
SELECT e.name
FROM public.employers e
LEFT JOIN public.employers_industries i on e.id = i.employer_id
GROUP BY e.id
HAVING count(i.industry_id) = 4
ORDER BY e.name
OFFSET 2 LIMIT 1
'''


In [80]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
print('Компания у которой заполнено 4 сферы деятельности и она находящаяся на 3 месте по алфавиту:', df_5_5.iloc [0][0])

Компания у которой заполнено 4 сферы деятельности и она находящаяся на 3 месте по алфавиту: 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [81]:
# текст запроса
query_5_6 = f'''
SELECT
COUNT(e.employer_id)
FROM public.employers_industries e
JOIN public.industries i on e.industry_id = i.id
WHERE(i.name) LIKE '%Разработка программного обеспечения%' 
'''

In [82]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
print('Разработка програмного обеспечения указана у', df_5_6.iloc [0][0], 'работодателей.')

Разработка програмного обеспечения указана у 3553 работодателей.


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [83]:
# адрес страницы c городами
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'

# сохранение списка городов 
cities = []

with requests.get(url) as response:
    soup = BeautifulSoup(response.text, 'html.parser')
    tables = soup.find_all('tbody')
    rows = tables[1].find_all('tr')
    
    for row in rows[1:]:
        cities.append(row.find('a')['title'])
t_cities= tuple(cities) #превращаем список в кортеж для передачи в запрос
print(t_cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [84]:
# текст запроса
query_5_7 = f'''
(SELECT 
    a.name as "Регион миллионник", 
    COUNT(v.id) as "Колличество вакансий"
FROM public.areas a
JOIN public.vacancies v ON a.id = v.area_id
JOIN public.employers e ON v.employer_id = e.id
WHERE e.name = 'Яндекс' AND a.name in {t_cities}
GROUP BY a.name)

UNION ALL

SELECT 'Total', count(*)
FROM public.vacancies v
 	JOIN public.employers e ON e.id = v.employer_id
 	JOIN public.areas a ON a.id = v.area_id
WHERE e.name = 'Яндекс' AND a.name in {t_cities} 
ORDER BY 2
'''

In [85]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
display(df_5_7)

Регион миллионник  Колличество вакансий
0               Омск                    21
1          Челябинск                    22
2         Красноярск                    23
3          Волгоград                    24
4              Пермь                    25
5             Казань                    25
6     Ростов-на-Дону                    25
7                Уфа                    26
8             Самара                    26
9          Краснодар                    30
10           Воронеж                    32
11       Новосибирск                    35
12   Нижний Новгород                    36
13      Екатеринбург                    39
14   Санкт-Петербург                    42
15            Москва                    54
16             Total                   485

***

## **Вывод:**

- Среди компаний, находящихся в исследуемой базе данных, наибольшее количество вакансий размещают IT компании, банки.

- Видно, что не все работадатели базы разместили какие-либо вакансиии.

- Часть крупных компаний размещают вакансии во многих регионах. Это связано как с предоставлением услуг в этих регионах, так и вероятно с желанием охватить более широкий круг людей, что увеличивает вероятность получения лучших специалистов в штат.

- Многие работадатели не указали сферу деятельности. Возможно в списке сфер деятельности не присутствует необходимых сфер или компания является широко известной в регионе размещения.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [86]:
# текст запроса
query_6_1 = f'''
SELECT 
    COUNT(v.name)
FROM public.vacancies v
WHERE lower(v.name) LIKE '%data%'
    OR lower(v.name) LIKE '%данн%' 
'''

In [87]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
print('Отношение к данным имеет', df_6_1.iloc [0][0], 'вакансий.')

Отношение к данным имеет 1771 вакансий.


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

В следующих заданиях мы продолжим работать с вакансиями по этому условию.

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [88]:
# текст запроса
query_6_2 = f'''
SELECT 
    COUNT(v.name)
FROM public.vacancies v
WHERE ((v.name ILIKE ANY(ARRAY['%data scientist%',
        '%data science%',
        '%исследователь данных%',
        '%machine learning%',
        '%машинн%обучен%'])
    OR v.name LIKE '%ML%') 
    AND v.name NOT ILIKE '%HTML%') 
    AND (v.name ILIKE '%junior%' 
    OR v.experience ILIKE '%нет опыта%'
    OR v.employment ILIKE '%стажировка%') 
'''

In [89]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
print('Количество подходящих вакансий для начинающего дата-сайентиста составляет:', df_6_2.iloc[0, 0])

Количество подходящих вакансий для начинающего дата-сайентиста составляет: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

* Критерии для отнесения вакансии к DS указаны в предыдущем задании.

In [90]:
# текст запроса
query_6_3 = f'''
SELECT 
    COUNT(v.name)
FROM public.vacancies v
WHERE ((v.name ILIKE ANY(ARRAY['%data scientist%',
        '%data science%',
        '%исследователь данных%',
        '%machine learning%',
        '%машинн%обучен%'])
    OR v.name LIKE '%ML%') 
    AND v.name NOT ILIKE '%HTML%')
    AND (v.key_skills ILIKE ANY(ARRAY['%sql%', '%postgres%']))
'''

In [91]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
print('Количество вакансий для DS c ключевыми навыками SQL или Рostgres:', df_6_3.iloc[0, 0])

Количество вакансий для DS c ключевыми навыками SQL или Рostgres: 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

* Это можно сделать помощью запроса, аналогичного предыдущему.

In [92]:
# текст запроса
query_6_4 = f'''
SELECT 
  COUNT(v.name)
FROM public.vacancies v
WHERE ((v.name ILIKE ANY(ARRAY['%data scientist%',
    '%data science%',
    '%исследователь данных%',
    '%machine learning%',
    '%машинн%обучен%'])
  OR v.name LIKE '%ML%') 
  AND v.name NOT ILIKE '%HTML%')
  AND (v.key_skills ILIKE ANY(ARRAY['%Python%']))  
'''

In [93]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
print('Количество вакансий для DS c ключевым навыком Python:', df_6_4.iloc[0, 0])

Количество вакансий для DS c ключевым навыком Python: 351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [94]:
# текст запроса
query_6_5 = f'''
SELECT 
    ROUND(avg(cardinality(regexp_split_to_array(key_skills, '\t'))), 2)
FROM public.vacancies v
WHERE ((v.name ILIKE ANY(ARRAY['%data scientist%',
        '%data science%',
        '%исследователь данных%',
        '%machine learning%',
        '%машинн%обучен%'])
    OR v.name LIKE '%ML%') 
    AND v.name NOT ILIKE '%HTML%')    
'''

In [95]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
print('В среднем для DS указывают', df_6_5.iloc [0][0],'ключевых навыков.')

В среднем для DS указывают 6.41 ключевых навыков.


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [96]:
# текст запроса
query_6_6 = f'''
SELECT
       DISTINCT v.experience as "Опыт работы",
       ROUND(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) as "Средняя зарплата"
FROM public.vacancies v
WHERE (LOWER(v.name) LIKE '%data scientist%' 
       OR LOWER(v.name) LIKE '%data science%' 
       OR LOWER(v.name) LIKE '%исследователь данных%' 
       OR (v.name LIKE '%ML%' and v.name NOT ILIKE '%HTML%') 
       OR LOWER(v.name) LIKE '%machine learning%' 
       OR LOWER(v.name) LIKE '%машинн%обучен%') 
       AND (v.salary_from IS NOT NULL or v.salary_to IS NOT NULL) 
       AND v.experience IS NOT NULL
GROUP BY 1 
'''

In [97]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
display(df_6_6)
print('В среднем дата-сайентист с опытом работы от 3 до 6 лет может рассчитывать на зарплату:', df_6_6.iloc [2][1])

Опыт работы  Средняя зарплата
0           Нет опыта           74643.0
1  От 1 года до 3 лет          139675.0
2       От 3 до 6 лет          243115.0

В среднем дата-сайентист с опытом работы от 3 до 6 лет может рассчитывать на зарплату: 243115.0


***

## **Вывод:**

* Отношение к данным имеет 1771 вакансия из 49197 представленных в БД.

* Количество подходящих вакансий для начинающего специалиста дата-сайентиста 51.

* В среднем для DS работодатели указывают 6 ключевых навыков.

* В представленной БД  "Python" как ключевой навык при приеме на работу отметили 351 работодатель.

* Дата-сайентист с опытом работы от 3 до 6 лет может рассчитывать на зарплату в среднем 243115.

* Дата-сайентист без опыта работы может рассчитывать на зарплату в среднем 74643.

In [98]:
connection.close()

# ***Общий вывод по проекту:***

* Анализ данных в представленном проэкте производился с помощью SQL-запросов, позволяющих подробно изучить рынок вакансий. Проект хорошо показывает, возможности по извлечению нужной информации хранящейся в базе данных.

* В целом, если говорить про рынок вакансий, то видно что наиболее высокая зарплата у дата-сейнтистов но войти в данную отрасль без опыта достаточно сложно поскольку для начинающих специалистов очень мало предложений, а для специалистов с опытом наоборот предложений в разы больше, и скорее всего там ощющается кадровый голод.

* Так же мы видим что предложения в основном сконцентрированы в самых больших городах что негативно сказывается на регионах где колличество вакансий вразы меньше а возможность удаленной работы не всегда предусмотрена что в свою очеред вызывает отток специалистов из маленьких регионов.
